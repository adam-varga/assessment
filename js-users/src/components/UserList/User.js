import React from "react";
import PropTypes from "prop-types";
import {
  Item,
  Info,
  ToggleUserStatusButton,
  Actions,
  EditButton,
  Name,
  CreatedAt,
} from "./UserList.styles";
import { connect } from "react-redux";
import { createAction, UPDATE_USER__REQUEST } from "../../store/actions";
import { USER_STATUSES, REQUEST_STATUSES } from "../../constants";
import { FiUnlock, FiLock, FiEdit2 } from "react-icons/fi";
import get from "lodash.get";

const statusButtonIconMap = {
  [USER_STATUSES.LOCKED]: FiLock,
  [USER_STATUSES.ACTIVE]: FiUnlock,
};

class User extends React.Component {
  render() {
    const { first_name, last_name, created_at, status, id } = this.props.user;

    const StatusButtonIcon = statusButtonIconMap[status];

    return (
      <Item>
        <Info {...this.props.user}>
          <Name>{`${first_name} ${last_name}`}</Name>
          <CreatedAt>{new Date(created_at).toLocaleDateString()}</CreatedAt>
        </Info>
        <Actions>
          <ToggleUserStatusButton
            data-testid={`toggle-user-status-button_${id}`}
            status={status}
            active={status === USER_STATUSES.LOCKED}
            onClick={this.props.toggleUserStatus}
            disabled={this.props.requestStatus === REQUEST_STATUSES.PENDING}
          >
            <StatusButtonIcon />
          </ToggleUserStatusButton>
          <EditButton
            to={`/edit/${id}`}
            data-testid={`edit-user-button_${id}`}
            disabled={this.props.requestStatus === REQUEST_STATUSES.PENDING}
          >
            <FiEdit2 />
          </EditButton>
        </Actions>
      </Item>
    );
  }
}

User.propTypes = {
  user: PropTypes.shape({
    first_name: PropTypes.string.isRequired,
    last_name: PropTypes.string.isRequired,
    status: PropTypes.oneOf(Object.values(USER_STATUSES)).isRequired,
  }).isRequired,
  requestStatus: PropTypes.oneOf(Object.values(REQUEST_STATUSES)),
  toggleUserStatus: PropTypes.func.isRequired,
};

function mapStateToProps(state) {
  return {
    requestStatus: get(state, `requests.${UPDATE_USER__REQUEST}.status`),
  };
}

function mapDispatchToProps(dispatch, ownProps) {
  return {
    toggleUserStatus: () =>
      dispatch(
        createAction(UPDATE_USER__REQUEST, {
          ...ownProps.user,
          status:
            ownProps.user.status === USER_STATUSES.LOCKED
              ? USER_STATUSES.ACTIVE
              : USER_STATUSES.LOCKED,
        })
      ),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(User);
