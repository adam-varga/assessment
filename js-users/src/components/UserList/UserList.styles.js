import styled, { css } from "styled-components";
import { USER_STATUSES } from "../../constants";
import { Button, ButtonLink } from "../common.styles";

export const Item = styled.li`
  padding: 5px;
  margin-bottom: 5px;
  list-style: none;
  display: flex;
`;

export const Info = styled.div`
  flex-grow: 1;
  ${(props) =>
    props.status === USER_STATUSES.LOCKED &&
    css`
      text-decoration: line-through;
      color: #aaa;
    `}
`;

export const Actions = styled.div`
  display: flex;
  align-items: center;
`;

export const EditButton = styled(ButtonLink)`
  margin-left: 10px;
`;

export const ToggleUserStatusButton = styled(Button)`
  ${(props) =>
    props.status === USER_STATUSES.LOCKED
      ? css`
          color: green;
          border-color: green;
        `
      : css`
          color: red;
          border-color: red;
        `}
`;

export const NewButton = styled(ButtonLink)`
  margin-bottom: 10px;
`;

export const List = styled.ul`
  padding: 0;
`;

export const Container = styled.div`
  .pagination {
    margin-top: 10px;
    li {
      list-style: none;
      display: inline-block;
      margin: 3px;
      a {
        display: flex;
        justify-content: center;
        align-items: center;
        min-width: 30px;
        padding: 0 5px;
        height: 30px;
        color: black;
        border: 1px solid transparent;
        cursor: pointer;
        &:focus {
          outline: none;
        }
      }
      &.active {
        a {
          border-color: black;
        }
      }
    }
  }
`;
