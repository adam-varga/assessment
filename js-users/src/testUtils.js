import { createStore } from "redux";
import React from "react";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import reducer from "./store/reducers";

export function renderWithRedux(
  ui,
  { initialState, store = createStore(reducer, initialState) } = {}
) {
  return {
    ...render(<Provider store={store}>{ui}</Provider>),
    store,
  };
}
