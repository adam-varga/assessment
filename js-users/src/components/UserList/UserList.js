import React from "react";
import { connect } from "react-redux";
import { createAction, GET_USERS__REQUEST } from "../../store/actions";
import { List, Container, NewButton, Actions, Title } from "./UserList.styles";
import User from "./User";
import ReactPaginate from "react-paginate";
import { USER_STATUSES, REQUEST_STATUSES } from "../../constants";
import {
  FiChevronLeft,
  FiChevronRight,
  FiMoreHorizontal,
  FiUserPlus,
} from "react-icons/fi";
import get from "lodash.get";
import PropTypes from "prop-types";

export class UserList extends React.Component {
  state = {
    startIndex: 0,
  };

  componentDidMount() {
    this.props.getUsers();
  }

  handlePageClick = (data) => {
    let selected = data.selected;
    let offset = Math.ceil(selected * this.props.pageSize);

    this.setState({ startIndex: offset });
  };

  render() {
    const { startIndex } = this.state;

    return (
      <React.Fragment>
        <Container>
          {!this.props.status ||
            (this.props.status === REQUEST_STATUSES.PENDING && "Loading...")}
          {this.props.users && (
            <React.Fragment>
              <Title>Users</Title>
              <Actions>
                <NewButton to="/new" data-testid="new-button">
                  <FiUserPlus />
                </NewButton>
              </Actions>
              <List>
                {this.props.users
                  .slice(startIndex, startIndex + this.props.pageSize)
                  .map((user) => (
                    <User key={user.id} user={user}></User>
                  ))}
              </List>
              <ReactPaginate
                previousLabel={<FiChevronLeft />}
                nextLabel={<FiChevronRight />}
                breakLabel={<FiMoreHorizontal />}
                pageCount={Math.ceil(
                  this.props.users.length / this.props.pageSize
                )}
                marginPagesDisplayed={1}
                pageRangeDisplayed={3}
                onPageChange={this.handlePageClick}
                containerClassName={"pagination"}
                subContainerClassName={"pages pagination"}
                activeClassName={"active"}
              />
            </React.Fragment>
          )}
        </Container>
      </React.Fragment>
    );
  }
}

UserList.defaultProps = {
  pageSize: 10,
};

UserList.propTypes = {
  users: PropTypes.arrayOf(
    PropTypes.shape({
      first_name: PropTypes.string.isRequired,
      last_name: PropTypes.string.isRequired,
      status: PropTypes.oneOf(Object.values(USER_STATUSES)).isRequired,
    })
  ),
  status: PropTypes.oneOf(Object.values(REQUEST_STATUSES)),
  getUsers: PropTypes.func.isRequired,
  pageSize: PropTypes.number,
};

function mapStateToProps(state) {
  return {
    status: get(state, `requests.${GET_USERS__REQUEST}.status`),
    users:
      state.users.userIds &&
      state.users.userIds.map((id) => state.users.usersById[id]),
  };
}

function mapDispatchToProps(dispatch) {
  return {
    getUsers: () => dispatch(createAction(GET_USERS__REQUEST)),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(UserList);
