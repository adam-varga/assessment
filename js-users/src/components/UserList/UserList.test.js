import React from "react";
import { cleanup, fireEvent } from "@testing-library/react";
import UserList from "./UserList";
import { renderWithRedux } from "../../testUtils";
import {
  createAction,
  GET_USERS__SUCCEEDED,
  CREATE_USER__SUCCEEDED,
  UPDATE_USER__SUCCEEDED,
} from "../../store/actions";
import { BrowserRouter as Router } from "react-router-dom";

afterEach(cleanup);

describe("User list", () => {
  let renderResult;

  beforeEach(() => {
    renderResult = renderWithRedux(
      <Router>
        <UserList pageSize={2} />
      </Router>
    );
  });

  it("renders loading screen by default", () => {
    const { getByText, asFragment } = renderResult;

    expect(getByText("Loading...")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  describe("after users request returned", () => {
    beforeEach(() => {
      const { store } = renderResult;

      store.dispatch(
        createAction(GET_USERS__SUCCEEDED, {
          usersById: {
            "1": {
              id: 1,
              last_name: "Person",
              first_name: "First",
              status: "active",
              created_at: "2014-01-10T08:28:49.030Z",
              updated_at: "2020-06-12T14:36:40.351Z",
              url: "http://1.json",
            },
            "2": {
              id: 2,
              last_name: "Person",
              first_name: "Second",
              status: "locked",
              created_at: "2014-01-10T08:28:49.059Z",
              updated_at: "2020-04-29T13:40:59.820Z",
              url: "http://2.json",
            },
            "3": {
              id: 3,
              last_name: "Person",
              first_name: "Third",
              status: "active",
              created_at: "2014-01-10T08:28:49.059Z",
              updated_at: "2020-04-29T13:40:59.820Z",
              url: "http://3.json",
            },
            "4": {
              id: 4,
              last_name: "Person",
              first_name: "Fourth",
              status: "active",
              created_at: "2014-01-10T08:28:49.030Z",
              updated_at: "2020-06-12T14:36:40.351Z",
              url: "http://4.json",
            },
          },
          userIds: ["1", "2", "3", "4"],
        })
      );
    });

    it("renders user list's first page", () => {
      const { queryByText, getByText, asFragment } = renderResult;

      expect(queryByText("Loading...")).toBeNull();

      expect(getByText("First Person")).toBeInTheDocument();
      expect(getByText("Second Person")).toBeInTheDocument();
      expect(queryByText("Third Person")).toBeNull();
      expect(queryByText("Fourth Person")).toBeNull();
      expect(asFragment()).toMatchSnapshot();
    });

    it("changes page after clicking page 2 button", () => {
      const { getByText, queryByText } = renderResult;

      const page2Button = getByText("2", { selector: "a" });

      fireEvent(
        page2Button,
        new MouseEvent("click", {
          bubbles: true,
          cancelable: true,
        })
      );

      expect(queryByText("First Person")).toBeNull();
      expect(queryByText("Second Person")).toBeNull();
      expect(getByText("Third Person")).toBeInTheDocument();
      expect(getByText("Fourth Person")).toBeInTheDocument();
    });

    it("changes URL after clicking New button", () => {
      const { getByTestId } = renderResult;
      const newButton = getByTestId("new-button");

      expect(global.location.pathname).toBe("/");

      fireEvent(
        newButton,
        new MouseEvent("click", {
          bubbles: true,
          cancelable: true,
        })
      );

      expect(global.location.pathname).toBe("/new");
    });

    it("adds newly added user as the first item of page 1", () => {
      const { getByText, queryByText, store } = renderResult;

      expect(getByText("First Person")).toBeInTheDocument();
      expect(getByText("Second Person")).toBeInTheDocument();
      expect(queryByText("New Person")).toBeNull();

      store.dispatch(
        createAction(CREATE_USER__SUCCEEDED, {
          user: {
            id: 5,
            last_name: "Person",
            first_name: "New",
            status: "active",
            created_at: "2014-01-10T08:28:49.030Z",
            updated_at: "2020-06-12T14:36:40.351Z",
            url: "http://5.json",
          },
        })
      );

      expect(getByText("First Person")).toBeInTheDocument();
      expect(queryByText("Second Person")).toBeNull();
      expect(getByText("New Person")).toBeInTheDocument();
    });

    it("updates user after update request successfully returned", () => {
      const { getByText, queryByText, store } = renderResult;

      expect(getByText("First Person")).toBeInTheDocument();
      expect(queryByText("Updated Person")).toBeNull();

      store.dispatch(
        createAction(UPDATE_USER__SUCCEEDED, {
          user: {
            id: 1,
            last_name: "Person",
            first_name: "Updated",
            status: "active",
            created_at: "2014-01-10T08:28:49.030Z",
            updated_at: "2020-06-12T14:36:40.351Z",
            url: "http://1.json",
          },
        })
      );

      expect(queryByText("First Person")).toBeNull();
      expect(getByText("Updated Person")).toBeInTheDocument();
    });
  });
});
