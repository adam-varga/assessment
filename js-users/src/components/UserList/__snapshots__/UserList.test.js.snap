// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User list after users request returned renders user list's first page 1`] = `
<DocumentFragment>
  <div
    class="sc-fzplWN bvzbOm"
  >
    <h1
      class="sc-AxiKw bcGqyg"
    >
      Users
    </h1>
    <div
      class="sc-AxheI kSIgOR"
    >
      <a
        class="sc-AxirZ sc-fzoLsD bJuYFa"
        data-testid="new-button"
        href="/new"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 21v-2a4 4 0 0 0-4-4H5a4 4 0 0 0-4 4v2"
          />
          <circle
            cx="8.5"
            cy="7"
            r="4"
          />
          <line
            x1="20"
            x2="20"
            y1="8"
            y2="14"
          />
          <line
            x1="23"
            x2="17"
            y1="11"
            y2="11"
          />
        </svg>
      </a>
    </div>
    <ul
      class="sc-fzpans kALogu"
    >
      <li
        class="sc-fzozJi jEWWTR"
      >
        <div
          class="sc-AxhCb kPkMwE"
          id="1"
        >
          <p
            class="sc-AxhUy bThcWD"
          >
            First Person
          </p>
          <p
            class="sc-AxgMl eyrKYH"
          >
            1/10/2014
          </p>
        </div>
        <div
          class="sc-AxheI kSIgOR"
        >
          <button
            class="sc-AxjAm sc-AxmLO gvIrbg"
            data-testid="toggle-user-status-button_1"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="11"
                rx="2"
                ry="2"
                width="18"
                x="3"
                y="11"
              />
              <path
                d="M7 11V7a5 5 0 0 1 9.9-1"
              />
            </svg>
          </button>
          <a
            class="sc-AxirZ sc-Axmtr eQgWKs"
            data-testid="edit-user-button_1"
            href="/edit/1"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M17 3a2.828 2.828 0 1 1 4 4L7.5 20.5 2 22l1.5-5.5L17 3z"
              />
            </svg>
          </a>
        </div>
      </li>
      <li
        class="sc-fzozJi jEWWTR"
      >
        <div
          class="sc-AxhCb bWMbKK"
          id="2"
        >
          <p
            class="sc-AxhUy bThcWD"
          >
            Second Person
          </p>
          <p
            class="sc-AxgMl eyrKYH"
          >
            1/10/2014
          </p>
        </div>
        <div
          class="sc-AxheI kSIgOR"
        >
          <button
            class="sc-AxjAm sc-AxmLO bgXSKk"
            data-testid="toggle-user-status-button_2"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="11"
                rx="2"
                ry="2"
                width="18"
                x="3"
                y="11"
              />
              <path
                d="M7 11V7a5 5 0 0 1 10 0v4"
              />
            </svg>
          </button>
          <a
            class="sc-AxirZ sc-Axmtr eQgWKs"
            data-testid="edit-user-button_2"
            href="/edit/2"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M17 3a2.828 2.828 0 1 1 4 4L7.5 20.5 2 22l1.5-5.5L17 3z"
              />
            </svg>
          </a>
        </div>
      </li>
    </ul>
    <ul
      class="pagination"
    >
      <li
        class="previous disabled"
      >
        <a
          aria-disabled="true"
          role="button"
          tabindex="0"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="15 18 9 12 15 6"
            />
          </svg>
        </a>
      </li>
      <li
        class="active"
      >
        <a
          aria-current="page"
          aria-label="Page 1 is your current page"
          role="button"
          tabindex="0"
        >
          1
        </a>
      </li>
      <li>
        <a
          aria-label="Page 2"
          role="button"
          tabindex="0"
        >
          2
        </a>
      </li>
      <li
        class="next"
      >
        <a
          aria-disabled="false"
          role="button"
          tabindex="0"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="9 18 15 12 9 6"
            />
          </svg>
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`User list renders loading screen by default 1`] = `
<DocumentFragment>
  <div
    class="sc-fzplWN bvzbOm"
  >
    Loading...
  </div>
</DocumentFragment>
`;
